File: leetcode/515-find-largest-value-in-each-tree-row/solution.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function largestValues(root: TreeNode | null): number[] {
  if (!root) {
    return [];
  }

  let queue: TreeNode[] = [root];
  const ans: number[] = [];

  while (queue.length) {
    const nextQueue: TreeNode[] = [];
    const nodesInCurrentLevel = queue.length;
    let maxVal: number = Number.NEGATIVE_INFINITY;

    for (let i = 0; i < nodesInCurrentLevel; i++) {
      const node = queue[i];

      maxVal = Math.max(maxVal, node.val);

      if (node.left) {
        nextQueue.push(node.left);
      }

      if (node.right) {
        nextQueue.push(node.right);
      }
    }

    queue = nextQueue;

    ans.push(maxVal);
  }

  return ans;
}

File: leetcode/515-find-largest-value-in-each-tree-row/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
vector<int> largestValues(TreeNode* root) {
    if (root == nullptr) {
        return {};
    }

    queue<TreeNode*> queue;
    queue.push(root);

    vector<int> ans;

    while (!queue.empty()) {
        int nodes_in_current_level = queue.size();
        int max_val = INT_MIN;

        for (int i = 0; i < nodes_in_current_level; i++) {
            TreeNode* node = queue.front();
            queue.pop();

            max_val = max(max_val, node->val);

            if (node->left) {
                queue.push(node->left);
            }

            if (node->right) {
                queue.push(node->right);
            }
        }

        ans.push_back(max_val);
    }

    return ans;
}
